### PR TITLE
Upgrading to tensorflow==2.7.0 - update interaction layer

### PR DESCRIPTION
**Pull request by lastmansleeping**
1. **https://github.com/salesforce/ml4ir/pull/156**
2. https://github.com/salesforce/ml4ir/pull/157
3. https://github.com/salesforce/ml4ir/pull/158
4. https://github.com/salesforce/ml4ir/pull/159

I was able to train both Ranking and EP models successfully with the following commands.
```
python ml4ir/applications/ranking/pipeline.py \
--data_dir ml4ir/applications/ranking/tests/data/tfrecord \
--feature_config ml4ir/applications/ranking/tests/data/configs/feature_config.yaml \
--run_id test \
--data_format tfrecord \
--loss_key softmax_cross_entropy \
--num_epochs 2 \
--execution_mode train_evaluate
```
and
```
python ml4ir/applications/classification/pipeline.py \
--data_dir ml4ir/applications/classification/tests/data/tfrecord \
--feature_config ml4ir/applications/classification/tests/data/configs/feature_config.yaml \
--model_config ml4ir/applications/classification/tests/data/configs/model_config.yaml \
--batch_size 32 \
--run_id test \
--data_format tfrecord \
--execution_mode train_evaluate \
--num_epochs 2
```

There are still a few pending items regardless and worth doing in follow up items before merging with master.

- [ ] Fix ranking *validation* metrics. Test metrics are working fine.
- [ ] Fix tests
- [ ] [FOLLOW UP] `model.summary()` doesn't print the nested layers, this causes issues with temperature scaling and L1 which require accessing weights from nested layers.
- [ ] [FOLLOW UP] Fix TF2.7 issue with handling duplicate metric names for old vs new metrics
- [ ] [FOLLOW UP]  Verify backward compatibility with `ml4ir-inference`

### CODE REVIEW

File: python/ml4ir/base/model/scoring/interaction_model.py [hunk 1]
```diff
@@ -9,93 +9,122 @@
...
-        train_features : dict
-            Dict of feature tensors that are used for training
-        metadata_features : dict
-            Dictionary of feature tensors that can be used for
-            computing custom metrics and losses
-        """
-        raise NotImplementedError
 
-    def transform_features_op(self, train_features, metadata_features):
+class UnivariateInteractionModel(InteractionModel):
```
Comment (arvindsrikantan) [inline python/ml4ir/base/model/scoring/interaction_model.py:77]: I see a duplicate below. Maybe you intended to inherit from Layer in the class below

Comment (arvindsrikantan) [reply python/ml4ir/base/model/scoring/interaction_model.py:77]: I just noticed that the InteractionModel inherits from Layer. Any reason why the class has been split up?

File: python/ml4ir/base/features/feature_fns/tf_native.py [hunk 2]
```diff
@@ -1,58 +1,77 @@
...
+        Tensor object
+            Modified feature tensor after applying all the specified ops
+        """
+
+        if not self.tf_ops:
+            return inputs
+
+        for tf_op in self.tf_ops:
+            try:
+                fn_, fn_args = eval(tf_op[self.FN]), tf_op.get(self.ARGS, {})
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/tf_native.py:62]: It's not really necessary to be done in `__init__` as these are just operations like `tf.math.add`, so it is not really a layer with a state worth initializing.

File: python/ml4ir/base/model/scoring/interaction_model.py (hunk 1 above)
Comment (lastmansleeping) [reply python/ml4ir/base/model/scoring/interaction_model.py:77]: Yes, I've cleaned this up in the follow up PR. Will move the change to this.

File: python/ml4ir/base/features/feature_fns/utils.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        if 0 in vocabulary_ids:
+            raise ValueError(
+                "Can not use ID 0 with dropout. Use categorical_embedding_with_vocabulary_file instead."
+            )
+    else:
+        vocabulary_ids = (
+            vocabulary_df[ID].values
+            if ID in vocabulary_df
+            else list(range(len(vocabulary_keys)))
+        )
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/utils.py:78]: I'm not sure I understand your question. The "dropout" here basically pushes some labels to the OOV index.

File: python/ml4ir/base/features/feature_fns/sequence.py [hunk 4]
```diff
@@ -4,221 +4,222 @@
...
-                pooled_tensors.append(tf.math.reduce_mean(masked_feature_tensor, axis=-1))
+                masked_feature_tensor = inputs
+
+            # Apply each pooling op
+            if fn == "sum":
+                pooled_tensors.append(tf.math.reduce_sum(masked_feature_tensor, axis=-1))
+
+            elif fn == "mean":
+                if "padded_val" in args:
+                    # NOTE: To avoid division by zero, we set 0 to a small value of 1e-10
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/sequence.py:182]: When the input is an empty list, it is padded to a fixed length.

File: python/ml4ir/base/features/feature_fns/categorical.py [hunk 5]
```diff
@@ -1,740 +1,554 @@
 import tensorflow as tf
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/categorical.py:1]: All the unit tests are passing for both feature_layer and preprocessing. This should be a good check IMO. I didn't have to modify any functional logic in the tests to get them to pass.

File: python/ml4ir/base/features/feature_fns/tf_native.py [hunk 6]
```diff
@@ -1,58 +1,77 @@
...
+                raise KeyError(
+                    "Invalid fn specified for tf_native_op : {}\n{}".format(tf_op[self.FN], e))
+
+            try:
+                feature_tensor = fn_(inputs, **fn_args)
+            except Exception as e:
+                raise Exception("Error while applying {} to {} feature:\n{}".format(
+                    tf_op[self.FN], self.feature_name, e))
+
+        # Adjusting the shape to the default feature fns for concatenating in the next step
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/tf_native.py:74]: The tl;dr explanation is the next step (concat to get a single tensor input for DNN) expects this shape. This is a limitation currently and we should find a cleaner way to do it. With that said, it is not a limitation that was introduced newly or something that could be solved easily.

File: python/ml4ir/base/features/feature_fns/base.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+from tensorflow.keras import layers
+
+from ml4ir.base.io.file_io import FileIO
+
+DEFAULT_VALUE = "default_value"
+
+
+class BaseFeatureLayerOp(layers.Layer):
+    """Abstract feature layer operation class"""
```
Comment (arvindsrikantan) [inline python/ml4ir/base/features/feature_fns/base.py:9]: One way to enforce `LAYER_NAME` in all subclasses is to have something like:
```python
    @property
    @abstractmethod
    def LAYER_NAME(self):
        raise AttributeError
```

in this class

File: python/ml4ir/base/features/feature_fns/sequence.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/base/features/feature_fns/sequence.py:206]: Okay. Case of missing features makes sense

File: python/ml4ir/base/features/feature_fns/utils.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    vocabulary_df = file_io.read_df(feature_layer_args[VOCABULARY_FILE])
+    if KEY in vocabulary_df.columns:
+        vocabulary_keys = vocabulary_df[KEY]
+    else:
+        vocabulary_keys = vocabulary_df.iloc[:, 0]
+    if MAX_LENGTH in feature_layer_args:
+        vocabulary_keys = vocabulary_keys[: feature_layer_args[MAX_LENGTH]]
+    if default_value or DEFAULT_VALUE in feature_layer_args:
+        default_value = default_value if default_value else feature_layer_args[DEFAULT_VALUE]
+        vocabulary_keys = vocabulary_keys.fillna(default_value)
```
[the remark this answers is not in the document]
Comment (arvindsrikantan) [reply python/ml4ir/base/features/feature_fns/utils.py:54]: @lastmansleeping , thoughts?

**Review by arvindsrikantan — COMMENTED** [review]
Can you switch off the tests we expect to be failing @lastmansleeping . It'll make it easier to ensure further development doesn't break valid tests

File: python/ml4ir/base/features/feature_fns/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_fns/utils.py:54]: I do not remember this.

**Comment by lastmansleeping** [comment]
> Can you switch off the tests we expect to be failing @lastmansleeping . It'll make it easier to ensure further development doesn't break valid tests

I will do this in the final (#158) PR.

**Review by arvindsrikantan — APPROVED** [review]

**Review by arvindsrikantan — APPROVED** [review]
